[OCPCLOUD-1702]: Update CAO to ignore platform related zone labels

odvarkadaniel opened the pull request:
This PR adds the ability for us to determine what provider platform are we using. With that knowledge, we can always append ignore zone labels for related platforms.
cc @elmiko

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -109,6 +110,17 @@ func AutoscalerArgs(ca *v1.ClusterAutoscaler, cfg *Config) []string {
+
+		switch strings.ToLower(cfg.CloudProvider) {
+		case "alibabacloud":
+			args = append(args, BalancingIgnoreLabelArg.Value("topology.diskplugin.csi.alibabacloud.com/zone"))
+		case "aws":
+			args = append(args, BalancingIgnoreLabelArg.Value("topology.ebs.csi.aws.com/zone"))
+		case "azure":
+			args = append(args, BalancingIgnoreLabelArg.Value("topology.disk.csi.azure.com/zone"))
+		case "gcp":
+			args = append(args, BalancingIgnoreLabelArg.Value("topology.gke.io/zone"))
```
Wonder if we should make the various cloud provider label names constants that we can group per platform? I feel like if we need to add more this could become confusing if they are embedded on individual lines

elmiko replied:
+1, i think that's a nice upgrade

elmiko reviewed (commented):
this is a good direction, i think we need to also bring the other aws and ibm labels that i referenced in this PR https://github.com/kubernetes/autoscaler/pull/5148

odvarkadaniel commented:
In the latest commit, I have added the rest of the labels that @elmiko pointed in his issue.
Also grouped the ignore labels per cloud provider and moved that into a function so we can work with that better in the future.

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -131,6 +219,8 @@ func AutoscalerArgs(ca *v1.ClusterAutoscaler, cfg *Config) []string {
 		args = append(args, VerbosityArg.Value(cfg.Verbosity))
 	}
 
+	fmt.Println(args)
```
Debug, remove before merge

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -68,6 +69,90 @@ const (
+// These values should be removed and replaced in the event that one of the cloud providers creates a nodegroupset processor.
+const (
+	// AlicloudIgnoredLabelCsiZone is a label used by the Alibaba Cloud CSI driver as a target for Persistent Volume Node Affinity.
+	AlicloudIgnoredLabelCsiZone = "topology.diskplugin.csi.alibabacloud.com/zone"
```
Do we want alibaba in their own const group?

odvarkadaniel replied:
Alibaba is mentioned in the PR Mike mentioned in earlier comment. What solution are you thinking about?

JoelSpeed replied:
Just all the other clouds are grouped individually, you have AWS consts, then azure, then GCP, but IBM and baba are mixed together?

odvarkadaniel replied:
Oh yes, missed that. Sorry. Fixed.

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -68,6 +69,90 @@ const (
+	GceIgnoredLabelGkeZone = "topology.gke.io/zone"
+)
+
+const (
```
Comment on the const group for consistency?

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -68,6 +69,93 @@ const (
+	AzureNodepoolLabel = "kubernetes.azure.com/agentpool"
+)
+
+const (
```
Comment for const group for consistency :)

JoelSpeed commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -68,6 +69,93 @@ const (
+	AwsIgnoredLabelEbsCsiZone = "topology.ebs.csi.aws.com/zone"
+)
+
+// The following values are for cloud providers which have not yet created specific nodegroupset processors.
+// These values should be removed and replaced in the event that one of the cloud providers creates a nodegroupset processor.
```
Do we want to make this comment consistent with the others?

odvarkadaniel replied:
Would you prefer to have this exact comment placed for GCP, Azure and Alibaba?

JoelSpeed replied:
Well, does this comment actually make sense in isolation? Maybe have it at the top, above all of the constant groups? As far as I know, none of these have specific nodegroupset processors because we are using them with the CAPI processor instead?

odvarkadaniel replied:
Yes, that is true. I will then place it at the top of the const declarations.

JoelSpeed reviewed (commented):
Couple of nits, otherwise LGTM

JoelSpeed commented:
/lgtm
/assign @elmiko for approve

elmiko commented on pkg/controller/clusterautoscaler/clusterautoscaler.go:
```diff
@@ -68,6 +69,95 @@ const (
+// AppendBasicIgnoreLabels appends ignore labels for specific cloud provider to the arguments
+// so the autoscaler can use these labels without the user having to input them manually.
+func appendBasicIgnoreLabels(args []string, cfg *Config) []string {
+	switch strings.ToLower(cfg.CloudProvider) {
```
i think it would be better if we could use the consts from here https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go to use for checking the platform type.

odvarkadaniel replied:
I think so as well. That is exactly what I proposed to you on Slack + some other things we could to do make our work easier in the future. I think it might be better for me to do that in separate PR on top of this change, wdyt?
I tried with this approach at first, but it turned out that there might be changes needed to some other parts of the CAO code base.

elmiko reviewed (commented):
mostly looks good to me, i think we should use the constant values for the platform types though.

also, were you going to refactor this @odvarkadaniel ?

elmiko reviewed (changes requested):
i don't think this approach will work the way it is. the cloud provider that is being used to detect the platform appears to be the value that is passed to the cluster-autoscaler binary as its `--cloud-provider` flag, which should always be `clusterapi`.

we need to detect the paltform type from the [platformType](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L60) in the infrastructure object, which means we will need to add the necessary rbac to read it, and then add a function to get the platform type from there.

see the following for current provider stuff:

default value
https://github.com/openshift/cluster-autoscaler-operator/blob/master/pkg/operator/config.go#L42-L44

setting default
https://github.com/openshift/cluster-autoscaler-operator/blob/master/pkg/operator/config.go#L143

trying to get from environment var
https://github.com/openshift/cluster-autoscaler-operator/blob/master/pkg/operator/config.go#L190

odvarkadaniel commented:
The latest commit retrieves the `providerType` from the `Infrastructure` object. We update the `clusterAutoscaler` object within the Reconcile function. We add a field that stores the `providerType` in the Autoscaler object.

JoelSpeed commented:
> We update the clusterAutoscaler object within the Reconcile function. We add a field that stores the providerType in the Autoscaler object.

I don't like the idea of storing this on the `ClusterAutoscaler` CRD, can we instead just look it up from the infrastructure object each time? We should be able to set up an informer for the infrastructure so that we have a cache.

The reason I don't like this is because the user could override this, and it creates a weird semantic for them, why do they care what platform it's on? It really should be internal to CAO IMO

JoelSpeed commented:
/lgtm

elmiko commented on pkg/controller/clusterautoscaler/clusterautoscaler_controller.go:
```diff
@@ -289,6 +293,13 @@ func (r *Reconciler) UpdateAutoscaler(ca *autoscalingv1.ClusterAutoscaler) error
+	if err != nil {
+		return err
+	}
+	r.config.platformType = platformType
```
do we need to set this on a `CreateAutoscaler` as well, or will we get an update after creating?

elmiko reviewed (commented):
generally looks good, i have a question and a suggestion.

elmiko reviewed (commented):
thanks @odvarkadaniel !

/approve
/lgtm

enxebre commented:
One of the main values of CAPI is that is cloud agnostic for consumers. Moving this cloud provider detail up here is a design regression in that sense. Let's try to not leak cloud provider details as we keep moving forward.

elmiko commented:
in general i agree with this notion @enxebre , but it seems like the cluster-api community would prefer to not encode these platform specific values into the cluster-autoscaler itself. it seems like the CAO is the right place to be adding these platform specific flags, do you think we should be pushing harder to have these values encoded into the autoscaler itself?
